### DISCUSSION TITLE
Command mvn package failed

### DISCUSSION THREAD

**Discussion by kkhanhluu** [keycloakify/keycloakify#138] [Q&A] [answered]
I'm currently running in ubuntu 20.04 and the mvn(Maven) is already installed by `apt install maven`. I got this error when i ran `npm run build && build-keycloak-theme`
![image](https://user-images.githubusercontent.com/38130967/179123302-f80b8fb5-2f45-47d0-bc2c-d9eb86bc3ef3.png)
Did someone also encounter this problem?

**Comment by garronej** [comment, marked as answer]
Hi @kkhanhluu 

Have you tried running `mvn package` in the `build_keycloak` directory?  

If yes, what does it says?

**Reply by kkhanhluu** [reply]
@garronej Thanks for your suggestion. I ran the command as you suggested and after reading this [answer](https://stackoverflow.com/a/67482132/12428566) I found out that, Ubuntu repackaged the maven, that means, I was not using Apache maven.

**Reply by garronej** [reply]
Thanks for taking the time to report the solution you found

**Comment by frodolfo** [comment]
@garronej 

I have `keycloak` in a Docker container.  I also get this error when I specify the following in `Dockerfile`
```
yarn build && ./node_modules/.bin/build-keycloak-theme
```
The error is thrown when it gets to the `build-keycloak-theme` executable,  The `child_process` (**line 90** of _src/bin/build-keycloak-theme/build-keycloak-theme.js_) isn't aware that `mvn` is already installed because it's a new shell, so presumably it  doesn't have the same PATH value, and therefore no reference to where mvn is.  Maven is invoked in other parts of Dockerfile with no issues.

Here's the output:
```
#25 498.4 /bin/sh: 1: mvn: not found
#25 498.4 child_process.js:866
#25 498.4     throw err;
#25 498.4     ^
#25 498.4
#25 498.4 Error: Command failed: mvn package
#25 498.4 /bin/sh: 1: mvn: not found
```

**Reply by garronej** [reply]
Hi @frodolfo,  
The `PATH` is forwarded to the child process.  
You getting this error either mean that:  
1) The user running `build-keycloak-theme` dosen't have `mvn` in the path.  
2) The user running `build-keycloak-theme` dosen't have +x right on `mvn`.  

I suggest you `RUN mvn --version` if I'm wrong you'll get something like:  

<img width="450" alt="image" src="https://user-images.githubusercontent.com/6702424/185545261-9585f342-9131-41f6-b74a-2fa4345b3801.png">

You can also try `yarn build && npx build-keycloak-theme`

**Reply by frodolfo** [reply]
@garronej  - I tried using `npx` like you suggested and it still didn't work

```
#25 124.6 🔏 Building the keycloak theme...⌚
#25 125.0 Not bundling email template because keycloak_email does not exist
#25 125.0 To start customizing the email template, run: 👉 npx create-keycloak-email-directory 👈
#25 167.9   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#25 167.9                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 34.8M  100 34.8M    0     0   837k      0  0:00:42  0:00:42 --:--:--  562k
#25 231.4   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#25 231.4                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 34.8M  100 34.8M    0     0   592k      0  0:01:00  0:01:00 --:--:--  567k
#25 233.5 /bin/sh: 1: mvn: not found
#25 233.5 child_process.js:866
```

Maven is available in my terminal and I execute mvn with no problems.  
```
mvn --version
Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)
```
I have a Makefile which runs `docker-compose` to create the Docker image.  The Dockerfile instructs docker-compose to execute `yarn && yarn build && npx build-keycloak-theme`, and it errs out when it gets to the child process invoking `execSync("mvn package")`.  I think when node is called in this virtual environment, it doesn't inherit the path value from the host terminal.  I need to figure out a way to propagate the path in the virtual environment.

**Reply by garronej** [reply]
@frodolfo,  
I am sorry you are experiencing thoses difficulties.  
I dont understand what's happening but I think your assumption aren't correct.  
> The command lookup is performed using the options.env.PATH environment variable if env is in the options object. Otherwise, `process.env.PATH` is used.  

[Ref](https://nodejs.org/api/child_process.html)  

[The call](https://github.com/InseeFrLab/keycloakify/blob/220d9d4d5eacbd839bf43ad274d344135a95142d/src/bin/build-keycloak-theme/build-keycloak-theme.ts#L90-L92)  

It's not that the `PATH` isn't fowarded to the `child_process`.  

> Maven is available in my terminal and I execute mvn with no problems.  

But is it in the context where you execute `yarn && yarn build && npx build-keycloak-theme`?  
Can you say for sure?

**Reply by frodolfo** [reply]
@garronej maybe I have to explicitly specify the PATH before calling `npx`.  Anyway, here's a recording of my terminal as I try to generate the Docker image.  I only included the relevant steps and cut the rest.  See MP4 file below:

https://drive.google.com/file/d/1aQbtSZsWUj4nLa28ipzHNgyeC6sbVO1O/view?usp=sharing

**Reply by garronej** [reply]
You'll have to give me access to you repo or create a minimal one if you want me to debug this.  
As stated before, if it's in the the `PATH` AND if the user running the process has `+x` privilege on `$(which mvn)` it should run...

**Comment by lordvlad** [comment]
you can check the most recent version of keycloakify, which will work without maven
